pkg/storage/utils/decomposedfs/upload.go

```diff
@@ -478,535 +262,50 @@ func (fs *Decomposedfs) PurgeExpiredUploads(purgedChan chan<- tusd.FileInfo) err
-			var targetEtag string
-			targetEtag, err = node.CalculateEtag(newNode.ID, fi.ModTime())
-			if err != nil {
-				return errtypes.InternalError(err.Error())
-			}
-			if ifMatch != targetEtag {
-				return errtypes.Aborted("etag mismatch")
-			}
-		}
-	} else {
-		// create dir to node
-		if err := os.MkdirAll(filepath.Dir(targetPath), 0700); err != nil {
-			sublog.Err(err).Msg("could not create node dir")
-			return errtypes.InternalError("could not create node dir")
-		}
-	}
-
-	// copy blob
-
-	file, err := os.Open(upload.binPath)
-	if err != nil {
-		return err
-	}
-	defer file.Close()
-	err = upload.fs.tp.WriteBlob(newNode, file)
-	if err != nil {
-		return errors.Wrap(err, "failed to upload file to blobstore")
-	}
-
-	// prepare discarding the blob if something changed while writing it
-	discardBlob := func() {
-		if err := upload.fs.tp.DeleteBlob(newNode); err != nil {
-			sublog.Err(err).Str("blobid", newNode.BlobID).Msg("Decomposedfs: failed to discard blob in blobstore")
-		}
-	}
-
-	// lock metadata node
-	lock, err := filelocks.AcquireWriteLock(targetPath)
-	if err != nil {
-		discardBlob()
-		return errtypes.InternalError(err.Error())
-	}
-	releaseLock := func() {
-		// ReleaseLock returns nil if already unlocked
-		if err := filelocks.ReleaseLock(lock); err != nil {
-			sublog.Err(err).Msg("Decomposedfs:could not unlock node")
-		}
-	}
-	defer releaseLock()
-
-	// check if match header again as safeguard
-	var oldMtime time.Time
-	versionsPath := ""
-	if fi, err = os.Stat(targetPath); err == nil {
-		// When the if-match header was set we need to check if the
-		// etag still matches before finishing the upload.
-		if ifMatch, ok := upload.info.MetaData["if-match"]; ok {
-			var targetEtag string
-			targetEtag, err = node.CalculateEtag(newNode.ID, fi.ModTime())
-			if err != nil {
-				discardBlob()
-				return errtypes.InternalError(err.Error())
-			}
-			if ifMatch != targetEtag {
-				discardBlob()
-				return errtypes.Aborted("etag mismatch")
-			}
-		}
-
-		// versions are stored alongside the actual file, so a rename can be efficient and does not cross storage / partition boundaries
-		versionsPath = upload.fs.lu.InternalPath(spaceID, newNode.ID+node.RevisionIDDelimiter+fi.ModTime().UTC().Format(time.RFC3339Nano))
-
-		// remember mtime of existing file so we can apply it to the version
-		oldMtime = fi.ModTime()
-	}
-
-	// read metadata
-
-	// attributes that will change
-	attrs := map[string]string{
-		xattrs.BlobIDAttr:   newNode.BlobID,
-		xattrs.BlobsizeAttr: strconv.FormatInt(newNode.Blobsize, 10),
-
-		// update checksums
-		xattrs.ChecksumPrefix + "sha1":    string(sha1h.Sum(nil)),
-		xattrs.ChecksumPrefix + "md5":     string(md5h.Sum(nil)),
-		xattrs.ChecksumPrefix + "adler32": string(adler32h.Sum(nil)),
-	}
-
-	// create version node with current metadata
-
-	var newMtime time.Time
-	// if file already exists
-	if versionsPath != "" {
-		// touch version node
-		file, err := os.Create(versionsPath)
```

Comment:
Instead of moving the existing node we create a new node. Thit fixes some nasty bugs, search for `// This move drops all metadata!!! We copy it below with CopyMetadata` in the old code which has moved to `updateExistingNode` in this PR.

You should use a `initNewRevisionNode`

Reply:
Got it. I'm only copying BlobID, BlobSize and Mtime from old node